### DISCUSSION TITLE
How to set BASE_URL environment variable in dockerized integration tests?

### DISCUSSION THREAD

**Discussion by Kolobamanacas** [vitest-dev/vitest#5695] [Q&A] [answered]
I'm using dockerized Nest.js application with integration tests written with Vitest. After upgrade of all the packages to the latest versions I face an issue I don't understand how to solve. My packages versions are:

```
@nestjs/core: 10.3.8
typescript:   5.4.5
vitest:       1.6.0
```

The issue I have is `TypeError: Invalid URL... Serialized Error: { code: 'ERR_INVALID_URL', input: '/' }` error thrown by `nestjs-url-generator`. It eventually happens because in tests a BASE_URL environment variable is set to `/` instead of `http://localhost` as it should be and as it set as an environment variable in `docker-compose.yml` file for the particular service.

Massive abuse of console.log() shows that during vitest's global setup, `process.env.BASE_URL` returns correct value `http://localhost`, but as soon as any testfile runs the `BASE_URL`'s  value is reset to `/`.

So it seems like Vite resets existing value of `BASE_URL`. I thought that if that's the case I could just reset it back in my `vitest-integration.config.ts` file as it described in the docs ([link1](https://vitejs.dev/guide/env-and-mode), [link2](https://vitejs.dev/config/shared-options#base)) like this:

```typescript
import { defineConfig, UserConfigExport } from 'vitest/config'

export default (): UserConfigExport => {
  return defineConfig({
    base: 'http://localhost', // <= I set new value for BASE_URL here.
  })
}
```

But that literally didn't change anything. So my question is: (best scenario) how do I force Vite to read existing value of `BASE_URL` from `docker-compose.yml`? Or, if that's not possible, how do I set any other value to `BASE_URL` other than default `/`? And lastly, if you also explain why in the world Vite replaces existing pre-defined environment variable I would really appreciate that. Thanks.

**Comment by hi-ogawa** [comment, marked as answer]
This seems like an unfortunate side effect from Vite. It looks like I can still overwrite it from `test.env` config. How about something like this?
https://stackblitz.com/edit/vitest-dev-vitest-f5v5k7?file=vite.config.ts

```ts
export default defineConfig({
  test: {
    env: {
      BASE_URL: process.env['BASE_URL'] ?? 'hello!',
    },
  },
});
```

It's certainly inconvenient but this `BASE_URL` convention seems deeply tied to Vite, so I'm not sure how much Vitest can manipulate that without breaking something.

**Reply by Kolobamanacas** [reply]
A bit hackish, but it works and seems like it didn't brake anything. Thanks! )
